# __slots__で軽量化
クラスの先頭で __slots__ を定義することで、クラスを軽量・高速にすることができます。

以下は、一般的なクラス定義です。

In [5]:
class Point(object):
     def __init__(self):
         self.x = 0
         self.y = 0

In [6]:
import sys
sys.getsizeof(Point)


1056

このクラスのサイズは 1056バイトとなりました

続いて、クラスの先頭で __slots__ を定義します。

__slots__ はリスト（or タプル）で、各要素は使用するアトリビュート名とします。

In [7]:
class Point(object):
     __slots__ = ['x', 'y']  # __slots__を定義
     def __init__(self):
         self.x = 0
         self.y = 0

In [8]:
sys.getsizeof(Point)

888

このクラスのサイズは 888バイトとなりました

---

## slotsの副作用

通常のクラスではアトリビュートを後から追加することができますが、

__slots__ を定義した場合には、__slots__ の要素のアトリビュートしか持てないようになります。

In [10]:
# __slot__を定義したクラスのインスタンス
p = Point()

# zを追加できない
p.z = 0

AttributeError: 'Point' object has no attribute 'z'